### PR TITLE
dfp: adding timing information about DNS resolution

### PR DESCRIPTION
**Pull request by alyssawilk**
Risk Level: low
Testing: new unit, integration tests
Docs Changes:
Release Notes: 
Part of https://github.com/envoyproxy/envoy-mobile/issues/1520

### CODE REVIEW

**Comment by alyssawilk** [comment]
Hey @mattklein123 pinging you for high level API thoughts.
The linked envoy mobile bug includes some other timing characteristics I'm going to add in follow-up PRs, namely time for the usptream connection to connect (start/ end) and time for SSL handshake (start/end)
right now I'm using the filterstate as it seems the simplest way to be extensible but if you have improvements to suggest I'd definitely prefer them earlier in this PR chain rather than later :-)

**Comment by mattklein123** [comment]
> Hey @mattklein123 pinging you for high level API thoughts.

At a high level I think using filter state is fine for this, but I have 2 questions/comments:
1) We should think about how to document this. Our filter state documentation is a bit haphazard and I want to make sure that people know about this.
2) Do we think there will be multiple related items that could be packaged together in a struct within a single key? I.e., instead of having a key per uint, etc. have a key per related data / struct that we could add on to? That might be a bit easier to understand.

**Comment by alyssawilk** [comment]
Agree this should have docs, but I wanted to make sure we agreed on where the info was stored, then sort out a doc location :-)

I like the idea of having a general struct.  What do we think of a hash map from timing string to ms_since_epoch or ms_since_start?   We could even TODO moving the backing store of the existing stream into metrics over to that and maybe even remove the timings APIs?  Then we can doc up a well known filter states doc parallel to well known dynamic metadata?  I think if we went that route I would lean towards adding it as a first class member to stream into APIs, but I'm good with whatever.

**Comment by mattklein123** [comment]
> I like the idea of having a general struct. What do we think of a hash map from timing string to ms_since_epoch or ms_since_start? We could even TODO moving the backing store of the existing stream into metrics over to that and maybe even remove the timings APIs? Then we can doc up a well known filter states doc parallel to well known dynamic metadata? I think if we went that route I would lean towards adding it as a first class member to stream into APIs, but I'm good with whatever.

Sure this sounds fine to me. The only thing I'm not sure of is whether we should key by string vs. just have some type of timing object that avoids hash lookups?

**Comment by alyssawilk** [comment]
if we want it to be extensible, what would you suggest other than hash map?  Given we're updating once per request/response I'm not worried about the hash lookup cost.
we already have an UpstreamTimingInfo struct.  I could instead move the downstream constructs to a DownstreamTimingInfo struct add extensible functions to both?

**Comment by mattklein123** [comment]
> we already have an UpstreamTimingInfo struct. I could instead move the downstream constructs to a DownstreamTimingInfo struct add extensible functions to both?

Ah OK got it. sure makes sense to me.

**Comment by junr03** [comment]
@alyssawilk do you want @mattklein123 to take a pass, or do you want me to now that you two discussed high level API concerns?

**Comment by alyssawilk** [comment]
yeah if you'd be up for a pass that'd be great

File: source/common/stream_info/stream_info_impl.h [hunk 1]
```diff
@@ -131,6 +131,13 @@ struct StreamInfoImpl : public StreamInfo {
     final_time_ = time_source_.monotonicTime();
   }
 
+  DownstreamTiming& downstreamTiming() override {
+    if (!downstream_timing_.has_value()) {
```
Comment (junr03) [inline source/common/stream_info/stream_info_impl.h:135]: Why have `downstream_timing_` as optional? Seems like we could just have the struct if we are constructing on absence?

File: source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc [hunk 2]
```diff
@@ -46,6 +54,8 @@ void ProxyFilter::onDestroy() {
 }
 
 Http::FilterHeadersStatus ProxyFilter::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
+  latchTime(decoder_callbacks_, dnsStart());
```
Comment (junr03) [inline source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc:57]: I'm curious about latching the time here vs. latching at other points in this function. For instance, only if the filter gets to `loadDnsCacheEntry`?

**Review by junr03 — COMMENTED** [review]
Thanks for working on this!

File: source/common/stream_info/stream_info_impl.h (hunk 1 above)
Comment (alyssawilk) [reply source/common/stream_info/stream_info_impl.h:135]: I could go either way here, but I don't think we set these for TCP, and TCP (so redis etc.) have stream info, so I was largely trying to spare them the memory footprint.

File: source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc (hunk 2 above)
Comment (alyssawilk) [reply source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc]: yeah it's the difference between when we start the lookup and when we start doing "DNS work"
I can go either way - moved.

File: source/common/stream_info/stream_info_impl.h (hunk 1 above)
Comment (junr03) [reply source/common/stream_info/stream_info_impl.h]: ah gotcha!

File: envoy/stream_info/stream_info.h [hunk 3]
```diff
@@ -240,6 +240,16 @@ struct UpstreamTiming {
 
 class DownstreamTiming {
 public:
+  void setValue(absl::string_view key, MonotonicTime value) { timings_[key] = value; }
```
Comment (mattklein123) [inline envoy/stream_info/stream_info.h:243]: Should we assert that we don't set a value twice? Not sure.

File: source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.h [hunk 4]
```diff
@@ -51,6 +51,13 @@ class ProxyFilter
 public:
   ProxyFilter(const ProxyFilterConfigSharedPtr& config) : config_(config) {}
 
+  static const std::string& dnsStart() {
```
Comment (mattklein123) [inline source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.h:54]: nit: I think you can just do `static constexpr absl::string_view = ...` for these.

**Review by mattklein123 — APPROVED** [review]
LGTM with small optional comments.

File: envoy/stream_info/stream_info.h (hunk 3 above)
Comment (alyssawilk) [reply envoy/stream_info/stream_info.h:243]: yeah I didn't know if folks using this would want to do updates, so left without.

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**